### DISCUSSION TITLE
Digital signature fails with PGPpainless, works with pgp cmd utility.

### DISCUSSION THREAD

**Discussion by elampros** [pgpainless/pgpainless#425] [Q&A] [answered]
I am definitely new with pgpainless, I am struggling with a sign first and then decrypt scenario. 

Here is the code

          ```
         setCustomSignatureHashPolicy(); for expiration dates

            SecretKeyRingProtector protector = PasswordBasedSecretKeyRingProtector.forKey(
                    factorSecretKey,
                    new Passphrase(passphrase.toCharArray())
            );

            SigningOptions signOpts = new SigningOptions();
            signOpts.addInlineSignature(protector, factorSecretKey,info.getPrimaryUserId(), DocumentSignatureType.BINARY_DOCUMENT);
            //signOpts.overrideHashAlgorithm(HashAlgorithm.SHA512);

            EncryptionOptions encOpts = EncryptionOptions.encryptCommunications();
            encOpts.addRecipient(bogPublicKeys);

            ProducerOptions options = ProducerOptions.signAndEncrypt(encOpts, signOpts);
            options.setAsciiArmor(false);
            options.overrideCompressionAlgorithm(CompressionAlgorithm.UNCOMPRESSED);

            ByteArrayOutputStream ciphertext = new ByteArrayOutputStream();
            EncryptionStream encryptionStream = PGPainless.encryptAndOrSign()
                    .onOutputStream(ciphertext)
                    .withOptions(options); // pass in the options object

            Streams.pipeAll(plaintextInputStream, encryptionStream); // pipe the data through
            encryptionStream.close(); // important! Close the stream to finish encryption/signing

            EncryptionResult result = encryptionStream.getResult(); // metadata
            return ciphertext.toByteArray();
```

The target system responds with Success on encryption and error digital signature verification.

When we send the file using pgp utility seems to work
First sign:

gpg -u GR01 --digest-algo SHA512 --compression-algo NONE --output TEST00001.zip.pgp  --passphrase p@ss --sign TEST00001.zip

then encrypt:

gpg -e -u GR01 -r "GRTEST_01 (Test Key for CR)" -o TEST00001.zip.pgp.pgp --compression-algo NONE TEST00001.zip.pgp

I am not sure what is wrong with the code. Can somebody help?

**Comment by vanitasvitae** [comment]
What software is running on the target system and what is the error message?

**Reply by elampros** [reply]
No idea, I am sending the file via a SOAP Web service. I guess Java, I really have no clue if they are using a library or spawning a pgp command. We started sending data via soapUI and now we need to implement a Java client. I reached this awesome repo, and I think I am close :-) 

They instruct us for inline signatures. I don't know how to proceed.

**Reply by elampros** [reply]
"Digital signature failed verification"

**Reply by vanitasvitae** [reply]
The way you describe creating an encrypted and signed message using gpg seems odd.
Usually, you do the signing and encryption in one step, e.g. `gpg --encrypt [...] --sign [...]`

I would guess, that doing it in two separate steps results in an unusual message format, so my guess would be that the target software expects this non-standard "double layer" message format?

You could try to replicate the same format by wrapping two EncryptionStreams, the inner doing inline-signing and the outer doing encryption, to verify this hypothesis. However, note that this should not be done in production.

**Reply by elampros** [reply]
It is really odd. It seems that we need to address this two step process. Can you help with the wrapping of the two EncryptionStreams? 
What you mean inner-outer? Can you please paste some pseudo/dummy code example?

**Reply by vanitasvitae** [reply, marked as answer]
Something along these lines:
```
//        setCustomSignatureHashPolicy(); for expiration dates

        SecretKeyRingProtector protector = PasswordBasedSecretKeyRingProtector.forKey(
                factorSecretKey,
                new Passphrase(passphrase.toCharArray())
        );

        SigningOptions signOpts = new SigningOptions();
        signOpts.addInlineSignature(protector, factorSecretKey,info.getPrimaryUserId(), DocumentSignatureType.BINARY_DOCUMENT);
        //signOpts.overrideHashAlgorithm(HashAlgorithm.SHA512);

        EncryptionOptions encOpts = EncryptionOptions.encryptCommunications();
        encOpts.addRecipient(bogPublicKeys);

        ProducerOptions innerOptions = ProducerOptions.sign(signOpts);
        innerOptions.setAsciiArmor(false);
        innerOptions.overrideCompressionAlgorithm(CompressionAlgorithm.UNCOMPRESSED);
        ProducerOptions outerOptions = ProducerOptions.encrypt(encOpts);
        outerOptions.setAsciiArmor(false);
        outerOptions.overrideCompressionAlgorithm(UNCOMPRESSED);

        // Outer stream for encryption
        ByteArrayOutputStream ciphertext = new ByteArrayOutputStream();
        EncryptionStream encryptionStream = PGPainless.encryptAndOrSign()
                .onOutputStream(ciphertext)
                .withOptions(outerOptions); // pass in the encryption options object

        // Inner stream for signing
        EncryptionStream signingStream = PGPainless.encryptAndOrSign()
                .onOutputStream(encryptionStream)
                .withOptions(innerOptions); // pass in signing options

        Streams.pipeAll(plaintextInputStream, signingStream);
        signingStream.close();
        encryptionStream.close();

        return ciphertext.toByteArray();
```

**Reply by elampros** [reply]
Oh my goodness! Worked Paul.

**Reply by vanitasvitae** [reply]
But please note: THIS IS NOT THE WAY THE OPENPGP PROTOCOL IS INTENDED TO BE USED!!!

You should get in touch with the vendor of the target software system and get them to fix their software :)

**Reply by elampros** [reply]
I spent the whole day trying several libraries. This one did the job. Thanks man.

Just a last question: You said to avoid the solution in production. The code looks pretty clean? What is the problem with those two stream in PROD?

**Reply by vanitasvitae** [reply]
First of all; It is non-standard and therefore causes interoperability problems as demonstrated by your bug report.
It isn't dangerous per say afaict, but I'd simply not do it. Doing signing and encryption in one step results in a clear message object, where it is obvoius how to procede with it (decrypt and then verify).

Having two separate steps results in an encrypted, yet not signed message, whos content is a binary file which the receiver has to guess is another OpenPGP message, which is unencrypted, yet signed.
Mostly: It is easy to avoid by using the protocol correctly.

**Reply by vanitasvitae** [reply]
I did some more digging and this is what I found:

Compared to the "combined method" (which you should use), the "two phase approach" suffers from a security issue called [surreptitious forwarding](https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html). This issue can be addressed with the combined method by the use of the (not yet very broadly adopted) "intended recipient subpacket", which ties information about the intended recipient of an encrypted message to the signed part of the message.

Further, the two phase approach is less efficient, since it requires two separate packet parser instances. With the combined method, one parser instance is sufficient and it is clear from the packet structure, how the message needs to be processed.
Also, depending on the implementation, there might be unnecessary format conversion happening between the different layers, but not necessarily.
